internal/cmd/oohelperd/quic.go

```diff
@@ -0,0 +1,94 @@
+package main
+
+//
+// QUIC connect measurements
+//
+
+import (
+	"context"
+	"crypto/tls"
+	"sync"
+	"time"
+
+	"github.com/lucas-clemente/quic-go"
+	"github.com/ooni/probe-cli/v3/internal/measurexlite"
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+)
+
+// ctrlQUICResult is the result of the QUIC check performed by the test helper.
+type ctrlQUICResult = model.THTLSHandshakeResult
+
+// quicResult contains the endpoint and the corresponding result.
+type quicResult struct {
+	// Address is the IP address we measured.
+	Address string
+
+	// Endpoint is the endpoint we measured.
+	Endpoint string
+
+	// QUIC contains the QUIC results
+	QUIC ctrlQUICResult
+}
+
+// quicConfig configures the QUIC connect check.
+type quicConfig struct {
+	// Address is the MANDATORY address to measure.
+	Address string
+
+	// Endpoint is the MANDATORY endpoint to connect to.
+	Endpoint string
+
+	// Logger is the MANDATORY logger to use.
+	Logger model.Logger
+
+	// NewQUICDialer is the MANDATORY factory for creating a new QUIC dialer.
+	NewQUICDialer func(model.Logger) model.QUICDialer
+
+	// Out is the MANDATORY chan where we'll post the QUIC measurement results.
+	Out chan *quicResult
+
+	// URLHostname is the MANDATORY URL.Hostname() to use.
+	URLHostname string
+
+	// Wg is MANDATORY and is used to sync with the parent.
+	Wg *sync.WaitGroup
+}
+
+// quicDo performs the QUIC handshake check.
+func quicDo(ctx context.Context, config *quicConfig) {
+	const timeout = 15 * time.Second
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	defer config.Wg.Done()
+	out := &quicResult{
+		Address:  config.Address,
+		Endpoint: config.Endpoint,
+		QUIC:     ctrlQUICResult{},
+	}
+	defer func() {
+		config.Out <- out
+	}()
+	ol := measurexlite.NewOperationLogger(
+		config.Logger,
+		"QUICConnect %s SNI=%s",
+		config.Endpoint,
+		config.URLHostname,
+	)
+	dialer := config.NewQUICDialer(config.Logger)
+	defer dialer.CloseIdleConnections()
+
+	tlsConfig := &tls.Config{
+		NextProtos: []string{"h3"},
+		RootCAs:    netxlite.NewDefaultCertPool(),
+		ServerName: config.URLHostname,
+	}
+	_, err := dialer.DialContext(ctx, config.Endpoint, tlsConfig, &quic.Config{})
```

Comment:
This code is leaking the returned `quic.EarlyConnection` and we must close it. I reckon that the `CloseIdleConnections` call above may be confusing in this regard. That call is about closing possibly idle connections used by resolvers used by the dialer and it has not impact on the dialed QUIC conn.

Do you think we should update the documentation to clarify the scope of `CloseIdleConnections`?

Here's what we need to do to fix the connection leak:

- [ ] define in `measurexlite` a function like `measurexlite.MaybeClose` for QUIC connections rather than for `net.Conn`, which should only close the connection if not nil;

- [ ] (obviously 😬) write tests for such a new function;

- [ ] make sure we use the function here;

- [ ] I see we're also leaking the connection in the `simplequicping` experiment, so I highly recommend that we also apply the fix of using the function there.